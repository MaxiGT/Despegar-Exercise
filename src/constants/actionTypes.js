export const GET_DELIVERY_SUCCESS = 'GET_DELIVERY_SUCCESS';
export const DELIVERY_CREATE = 'CREATE_DELIVERY';
export const DELIVERY_DELETE = 'DELETE_DELIVERY';
export const DELIVERY_UPDATE = 'EDIT_DELIVERY';
export const DELIVERY_SELECT = 'SELECT_DELIVERY';
export const DELIVERY_RECOVER = 'DELIVERY_RECOVER';
export const SET_ORDER = 'SET_ORDER';

export const CHANGE_FILTER = 'CHANGE_FILTER';
export const RAISE_ERROR = 'RAISE_ERROR';
export const SIMULATE_API_CALL = 'SIMULATE_API_CALL';