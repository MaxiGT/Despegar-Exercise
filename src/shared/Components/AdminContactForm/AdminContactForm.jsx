import React from 'react';

class AdminContactForm extends React.Component {
  constructor(props) {
    super(props);
    const { value = {}, isAdmin } = this.props;
    this.state = {
      ...value,
      isAdmin,
    };
  }

  handleChange = (evt) => {
    this.setState({
      [evt.target.name]: evt.target.value,
    }, () => {
      this.props.onChange(this.state);
    });
  }

  render() {

    const { config } = this.props;
    
    return (
      <div style={{ paddingLeft: '0' }} className='col-9'>
        {
          config.map((i, idx) => {
            const Component = i.component;
            return (
              <Component key={idx} onChange={this.handleChange} value={this.state[i.name]} {...i} />
            );
          })
        }
      </div>
    );
  }
}

export default AdminContactForm;