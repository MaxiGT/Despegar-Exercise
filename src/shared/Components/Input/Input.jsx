import React from  'react';
import PropTypes from 'prop-types';

class Input extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      errorMessage: ''
    };
  }
  
  handleOnChange = (evt) => {
    if (this.props.onChange) this.props.onChange(evt);
    if (this.state.errorMessage !== '') {
      const errorMessage = '';
      this.setState({
        errorMessage,
      });
    }
  }

  handleBlur = (evt) => {
    if (this.props.onBlur) this.props.onBlur(evt);
    if (this.props.isRequired && evt.target.value === '') {
      const errorMessage = 'Este campo es obligatorio';
      this.setState({
        errorMessage,
      });
    }
  }

  render() {
    const { name, value, label, placeholder, type, length, append } = this.props;
  
    return (
      <div className='form-group'>
        <div className="input-group">
          {
            append && (
              <div className="input-group-prepend">
                <span className="input-group-text" id="inputGroup-sizing-default">{label}</span>
              </div>
            )
          }
          <input
            type={type}
            name={name}
            value={value}
            maxLength={length}
            onChange={(evt) => this.handleOnChange(evt)}
            onBlur={(evt) => this.handleBlur(evt)}
            className="form-control"
            placeholder={placeholder}
            aria-label="Default"
            aria-describedby="inputGroup-sizing-default" />
            {
            !append && (
              <div className="input-group-prepend">
                <span className="input-group-text" id="inputGroup-sizing-default">{label}</span>
              </div>
            )
          }
        </div>
        <span className='help-block float-right'
          style={{ color: 'red', fontWeight: 'bold' }}>{this.state.errorMessage}</span>
      </div>
    );
  }
}

Input.propTypes = {
  type: PropTypes.string,
  name: PropTypes.string,
  placeholder: PropTypes.string,
  label: PropTypes.string,
  onChange: PropTypes.func,
  append: PropTypes.bool,
}

Input.defaultProps = {
  type: 'text',
  placeholder: '',
  append: true,
}

export default Input;