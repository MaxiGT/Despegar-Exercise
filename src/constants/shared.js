import Input from '../shared/Components/Input/Input';
import TextArea from '../shared/Components/TextArea/TextArea';

export const filtersConfig = [
  {
    isSelected: true,
    label: 'Nombre',
    name: 'name',
    type: 'text',
    placeholder: 'Filtre por Nombre',
  },
  {
    isSelected: true,
    label: 'Direccion',
    name: 'address',
    type: 'text',
    placeholder: 'Filtre por direccion',
  },
  {
    isSelected: false,
    label: 'Telefono',
    name: 'phone',
    type: 'text',
    placeholder: 'Filtre por telefono',
  },
];

export const inputsForm = [
  {
    label: 'Nombre',
    name: 'name',
    isRequired: true,
    length: 50,
    component: Input,
    validateField: ['name'],
  },
  {
    label: 'Telefono',
    name: 'phoneNumber',
    type: 'tel',
    isRequired: false,
    length: 20,
    component: Input,
  },
  {
    label: 'Descripcion',
    name: 'description',
    isRequired: true,
    length: 1000,
    component: TextArea,
    validateField: ['description'],
  },
  {
    label: 'Especialidades',
    name: 'specialty',
    isRequired: false,
    length: 500,
    component: TextArea,
  },
  {
    label: 'Direccion',
    name: 'address',
    isRequired: true,
    length: 50,
    component: Input,
    validateField: ['address'],
  },
];

export const adminContactInputs = [
  {
    label: 'Nombre',
    name: 'name',
    isRequired: true,
    length: 200,
    component: Input,
    validateField: ['adminContact', 'name'],
  },
  {
    label: 'Apellido',
    name: 'lastName',
    isRequired: true,
    length: 200,
    component: Input,
    validateField: ['adminContact', 'lastName'],
  },
  {
    label: 'Telefono',
    name: 'phone',
    isRequired: true,
    length: 100,
    component: Input,
    validateField: ['adminContact', 'phone'],
  },
  {
    label: 'Email',
    name: 'email',
    isRequired: true,
    length: 50,
    component: Input,
    validateField: ['adminContact', 'email'],
  },
];

export const comercialContactInputs = [
  {
    label: 'Nombre',
    name: 'name',
    isRequired: true,
    length: 200,
    component: Input,
    validateField: ['comercialContact', 'name'],
  },
  {
    label: 'Apellido',
    name: 'lastName',
    isRequired: true,
    length: 200,
    component: Input,
    validateField: ['comercialContact', 'lastName'],
  },
  {
    label: 'Telefono',
    name: 'phone',
    isRequired: true,
    length: 100,
    component: Input,
    validateField: ['comercialContact', 'phone'],
  },
  {
    label: 'Email',
    name: 'email',
    isRequired: true,
    length: 50,
    component: Input,
    validateField: ['comercialContact', 'email'],
  },
];