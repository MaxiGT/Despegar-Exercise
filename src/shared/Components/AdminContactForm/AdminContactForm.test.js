import React from 'react';
import { adminContactInputs } from '../../../constants/shared';
import { shallow } from 'enzyme';
import AdminContactForm from './AdminContactForm';
import { spy } from 'sinon';
import Input from '../Input/Input';

describe('AdminContactForm', () => {

  const onChangeMock = spy();
  const isAdmin = true;
  const initialValue = {
    name: 'Admin Mock Name',
    lastName: 'Mock LastName',
    phone: '456 715 42',
    email: 'MockEmail@MockEmail.com',
  };
  const newValue = {
    name: 'New Admin Mock Name',
    lastName: 'New Mock LastName',
    phone: '(011) 456 715 42',
    email: 'NewMockEmail@MockEmail.com',
  };

  const form = shallow(<AdminContactForm
    config={adminContactInputs}
    onChange={onChangeMock}
    isAdmin={isAdmin}
    value={initialValue} />);

  it('Should render properly', () => {
    expect(form.find(Input).length).toEqual(adminContactInputs.length);
    expect(form).toMatchSnapshot();
  });

  it('Should react input change', () => {
    form.find({ name: 'name' }).simulate('change', { target: { name: '', value: '' } });
    expect(onChangeMock.called).toBe(true);
  });

  it('Should react to props change', () => {
    expect(form.instance().props.value).toEqual(initialValue);
    form.setProps({ value: newValue });
    expect(form.instance().props.value).toEqual(newValue);
  });
});

