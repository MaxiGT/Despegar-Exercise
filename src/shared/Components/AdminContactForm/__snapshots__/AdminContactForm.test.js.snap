// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdminContactForm Should render properly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AdminContactForm
    config={
      Array [
        Object {
          "component": [Function],
          "isRequired": true,
          "label": "Nombre",
          "length": 200,
          "name": "name",
          "validateField": Array [
            "adminContact",
            "name",
          ],
        },
        Object {
          "component": [Function],
          "isRequired": true,
          "label": "Apellido",
          "length": 200,
          "name": "lastName",
          "validateField": Array [
            "adminContact",
            "lastName",
          ],
        },
        Object {
          "component": [Function],
          "isRequired": true,
          "label": "Telefono",
          "length": 100,
          "name": "phone",
          "validateField": Array [
            "adminContact",
            "phone",
          ],
        },
        Object {
          "component": [Function],
          "isRequired": true,
          "label": "Email",
          "length": 50,
          "name": "email",
          "validateField": Array [
            "adminContact",
            "email",
          ],
        },
      ]
    }
    isAdmin={true}
    onChange={[Function]}
    value={
      Object {
        "email": "MockEmail@MockEmail.com",
        "lastName": "Mock LastName",
        "name": "Admin Mock Name",
        "phone": "456 715 42",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Input
          append={true}
          component={[Function]}
          isRequired={true}
          label="Nombre"
          length={200}
          name="name"
          onChange={[Function]}
          placeholder=""
          type="text"
          validateField={
            Array [
              "adminContact",
              "name",
            ]
          }
          value="Admin Mock Name"
        />,
        <Input
          append={true}
          component={[Function]}
          isRequired={true}
          label="Apellido"
          length={200}
          name="lastName"
          onChange={[Function]}
          placeholder=""
          type="text"
          validateField={
            Array [
              "adminContact",
              "lastName",
            ]
          }
          value="Mock LastName"
        />,
        <Input
          append={true}
          component={[Function]}
          isRequired={true}
          label="Telefono"
          length={100}
          name="phone"
          onChange={[Function]}
          placeholder=""
          type="text"
          validateField={
            Array [
              "adminContact",
              "phone",
            ]
          }
          value="456 715 42"
        />,
        <Input
          append={true}
          component={[Function]}
          isRequired={true}
          label="Email"
          length={50}
          name="email"
          onChange={[Function]}
          placeholder=""
          type="text"
          validateField={
            Array [
              "adminContact",
              "email",
            ]
          }
          value="MockEmail@MockEmail.com"
        />,
      ],
      "className": "col-9",
      "style": Object {
        "paddingLeft": "0",
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "0",
        "nodeType": "class",
        "props": Object {
          "append": true,
          "component": [Function],
          "isRequired": true,
          "label": "Nombre",
          "length": 200,
          "name": "name",
          "onChange": [Function],
          "placeholder": "",
          "type": "text",
          "validateField": Array [
            "adminContact",
            "name",
          ],
          "value": "Admin Mock Name",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "1",
        "nodeType": "class",
        "props": Object {
          "append": true,
          "component": [Function],
          "isRequired": true,
          "label": "Apellido",
          "length": 200,
          "name": "lastName",
          "onChange": [Function],
          "placeholder": "",
          "type": "text",
          "validateField": Array [
            "adminContact",
            "lastName",
          ],
          "value": "Mock LastName",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "2",
        "nodeType": "class",
        "props": Object {
          "append": true,
          "component": [Function],
          "isRequired": true,
          "label": "Telefono",
          "length": 100,
          "name": "phone",
          "onChange": [Function],
          "placeholder": "",
          "type": "text",
          "validateField": Array [
            "adminContact",
            "phone",
          ],
          "value": "456 715 42",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "3",
        "nodeType": "class",
        "props": Object {
          "append": true,
          "component": [Function],
          "isRequired": true,
          "label": "Email",
          "length": 50,
          "name": "email",
          "onChange": [Function],
          "placeholder": "",
          "type": "text",
          "validateField": Array [
            "adminContact",
            "email",
          ],
          "value": "MockEmail@MockEmail.com",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Input
            append={true}
            component={[Function]}
            isRequired={true}
            label="Nombre"
            length={200}
            name="name"
            onChange={[Function]}
            placeholder=""
            type="text"
            validateField={
              Array [
                "adminContact",
                "name",
              ]
            }
            value="Admin Mock Name"
          />,
          <Input
            append={true}
            component={[Function]}
            isRequired={true}
            label="Apellido"
            length={200}
            name="lastName"
            onChange={[Function]}
            placeholder=""
            type="text"
            validateField={
              Array [
                "adminContact",
                "lastName",
              ]
            }
            value="Mock LastName"
          />,
          <Input
            append={true}
            component={[Function]}
            isRequired={true}
            label="Telefono"
            length={100}
            name="phone"
            onChange={[Function]}
            placeholder=""
            type="text"
            validateField={
              Array [
                "adminContact",
                "phone",
              ]
            }
            value="456 715 42"
          />,
          <Input
            append={true}
            component={[Function]}
            isRequired={true}
            label="Email"
            length={50}
            name="email"
            onChange={[Function]}
            placeholder=""
            type="text"
            validateField={
              Array [
                "adminContact",
                "email",
              ]
            }
            value="MockEmail@MockEmail.com"
          />,
        ],
        "className": "col-9",
        "style": Object {
          "paddingLeft": "0",
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "0",
          "nodeType": "class",
          "props": Object {
            "append": true,
            "component": [Function],
            "isRequired": true,
            "label": "Nombre",
            "length": 200,
            "name": "name",
            "onChange": [Function],
            "placeholder": "",
            "type": "text",
            "validateField": Array [
              "adminContact",
              "name",
            ],
            "value": "Admin Mock Name",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "class",
          "props": Object {
            "append": true,
            "component": [Function],
            "isRequired": true,
            "label": "Apellido",
            "length": 200,
            "name": "lastName",
            "onChange": [Function],
            "placeholder": "",
            "type": "text",
            "validateField": Array [
              "adminContact",
              "lastName",
            ],
            "value": "Mock LastName",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "2",
          "nodeType": "class",
          "props": Object {
            "append": true,
            "component": [Function],
            "isRequired": true,
            "label": "Telefono",
            "length": 100,
            "name": "phone",
            "onChange": [Function],
            "placeholder": "",
            "type": "text",
            "validateField": Array [
              "adminContact",
              "phone",
            ],
            "value": "456 715 42",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "3",
          "nodeType": "class",
          "props": Object {
            "append": true,
            "component": [Function],
            "isRequired": true,
            "label": "Email",
            "length": 50,
            "name": "email",
            "onChange": [Function],
            "placeholder": "",
            "type": "text",
            "validateField": Array [
              "adminContact",
              "email",
            ],
            "value": "MockEmail@MockEmail.com",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
